Detect Browser Autofill

jdnichollsc opened the pull request:
Fix onIonChange is not detecting browser autoFill event from React

<!-- Please refer to our contributing documentation for any questions on submitting a pull request, or let us know here if you need any help: https://ionicframework.com/docs/building/contributing -->

## Pull request checklist

Please check if your PR fulfills the following requirements:
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [ ] Lint (`npm run lint`) has passed locally and any fixes were made for failures


## Pull request type

<!-- Please do not submit updates to dependencies unless it fixes an issue. --> 

<!-- Please try to limit your pull request to one type, submit multiple pull requests if needed. --> 

Please check the type of change your PR introduces:
- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: https://github.com/ionic-team/ionic-framework/issues/21466


## What is the new behavior?
<!-- Please describe the behavior or changes that are being added by this PR. -->
Fix wrong behavior with autofill

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

<!-- If this introduces a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

Please let me know what you think 👍

liamdebeasi commented:
Thanks for the PR! Could you please provide some steps to reproduce the issue that this fixes? Looks like the original issue was closed because I could not reproduce the issue.

jdnichollsc commented:
Oh ok, I have an example with Ionic React here https://github.com/jdnichollsc/adventure-closure/blob/master/app/src/components/SignIn/index.tsx#L91
I added autocomplete='off' because the React state is not updated when Chrome fill these fields

What do you think?

liamdebeasi commented:
Could you clarify how the issue manifests? I am testing this in a blank Ionic starter app and when Chrome autofills a value (I am testing with a street address), `onIonChange` fires properly with the correct value.

jdnichollsc commented:
Ohh very odd, can you share that from [StackBlitz](https://stackblitz.com/) please?

liamdebeasi commented:
Here's an example: https://stackblitz.com/edit/ionic-react-autofill

Feel free to fork and make changes if I am doing something wrong.

jdnichollsc commented:
Thanks Liam for your help! <3
I updated your example to detect the issue when browser autofill the fields of the login page when it's mounted: https://stackblitz.com/edit/ionic-react-autofill-eczb9h?file=index.js

Also I updated my example here: https://github.com/jdnichollsc/adventure-closure/blob/master/app/src/components/SignIn/index.tsx#L109
This is the link to reproduce the issue with my example: https://adventure-closure.web.app/login
And please check the below video: 

https://user-images.githubusercontent.com/2154886/113370073-92744380-9328-11eb-8ce5-3933eea4c4be.mp4

Let me know what you think 👍

liamdebeasi commented:
Thanks! While I can reproduce the behavior you are describing, I think this is the intended behavior. It looks as though Chrome does not update the internal value of an input until you interact with a page.

I was able to reproduce the same behavior outside of Ionic Framework: 

```html
<form autocomplete="on">
  <input id="email" type="text" autocomplete="email" placeholder="email" />
  <input type="password" autocomplete="current-password" name="password" placeholder="password" />
  <button type="submit">Submit</button>
</form>
```

I added a `setTimeout` to execute after `5000ms`, reloaded the page, and then didn't do anything. Logging the value of the inputs in the `setTimeout` callback showed that Chrome did not update the value of the input. However, if I added a `change` event listener to the input and interacted with the page, then Chrome would then update the internal value.

Is this in line with the behavior you are seeing too?

jdnichollsc commented:
Exactly! Very odd 🤔

liamdebeasi commented:
I agree it is strange. I tried to find some documentation surrounding this, but maybe I was not looking in the right location. I am going to close this for now since it seems like the current Chrome behavior is the intended behavior.
